### PR TITLE
Stabilize 128-bit integers :tada:

### PR DESCRIPTION
**Pull request by mark-i-m**
cc #35118 

EDIT: This should be merged only after the following have been merged:
- [x] https://github.com/rust-lang-nursery/compiler-builtins/pull/236
- [x] https://github.com/rust-lang/book/pull/1230

### CODE REVIEW

**Comment by mark-i-m** [comment]
What does this tidy error mean?

```
tidy error: /checkout/src/libcore/num/mod.rs:3961: mismatches to previous in: ["stability level", "since", "tracking issue"]
```

**Comment by mark-i-m** [comment]
I'm also getting this error:

```
   Compiling std_unicode v0.0.0 (file:///nobackup/rust/src/libstd_unicode)
error: this feature has been stable since 1.26.0. Attribute no longer needed
  --> rustc/compiler_builtins_shim/../../libcompiler_builtins/src/lib.rs:15:12
   |
15 | #![feature(i128_type)]
   |            ^^^^^^^^^
   |
note: lint level defined here
  --> rustc/compiler_builtins_shim/../../libcompiler_builtins/src/lib.rs:1:31
   |
1  | #![cfg_attr(not(stage0), deny(warnings))]
   |                               ^^^^^^^^
   = note: #[deny(stable_features)] implied by #[deny(warnings)]

error: aborting due to previous error

error: Could not compile `compiler_builtins`.
```

**Comment by PlasmaPower** [comment]
For the latter error, it looks like `libcompiler_builtins/src/lib.rs` has `#![feature(i128_type)]` on line 15 (because it was unstable). Now that it's stable, it's erroring, because it also has `#![deny(warnings)]` and the compiler warns when you `#![feature()]` an already stable feature.

**Comment by PlasmaPower** [comment]
Here's the line: https://github.com/rust-lang-nursery/compiler-builtins/blob/266ea0740a5bdd262a38bbd88fb55fc3d2a7a96e/src/lib.rs#L15

Unfortunately it's in a submodule (different repository) so I don't think there's a good way to simultaneously change them.

File: src/doc/unstable-book/src/language-features/repr128.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# `repri128`
```
Comment (goodmanjonathan) [inline src/doc/unstable-book/src/language-features/repr128.md]: s/repri128/repr128

File: src/libsyntax/diagnostic_list.rs [hunk 2]
```diff
@@ -261,10 +264,11 @@ If you're using a nightly version of rustc, just add the corresponding feature
 to be able to use it:
 
 ```
-#![feature(i128)]
+#![feature(repri128)]
```
Comment (goodmanjonathan) [inline src/libsyntax/diagnostic_list.rs]: s/repri128/repr128

File: src/libsyntax/diagnostic_list.rs [hunk 3]
```diff
@@ -250,7 +250,10 @@ An unstable feature was used.
 Erroneous code example:
 
 ```compile_fail,E658
-let x = ::std::u128::MAX; // error: use of unstable library feature 'i128'
+#[repr(u128)] // error: use of unstable library feature 'i128'
```
Comment (goodmanjonathan) [inline src/libsyntax/diagnostic_list.rs]: nit: language feature `repr128`

File: src/test/ui/error-codes/E0658.stderr [hunk 4]
```diff
@@ -4,7 +4,7 @@ error[E0658]: use of unstable library feature 'i128' (see issue #35118)
 LL |     let _ = ::std::u128::MAX; //~ ERROR E0658
    |             ^^^^^^^^^^^^^^^^
    |
-   = help: add #![feature(i128)] to the crate attributes to enable
+   = help: add #![feature(repri128)] to the crate attributes to enable
```
Comment (goodmanjonathan) [inline src/test/ui/error-codes/E0658.stderr]: s/repri128/repr128

File: src/test/ui/feature-gate-i128_type2.rs [hunk 5]
```diff
@@ -10,20 +10,20 @@
 
 // gate-test-i128_type
```
Comment (goodmanjonathan) [inline src/test/ui/feature-gate-i128_type2.rs]: Why not remove the `i128_type` tests?

File: src/test/run-pass/saturating-float-casts.rs [hunk 6]
```diff
@@ -11,7 +11,7 @@
 // Tests saturating float->int casts. See u128-as-f32.rs for the opposite direction.
 // compile-flags: -Z saturating-float-casts
 
-#![feature(test, i128, i128_type, stmt_expr_attributes)]
+#![feature(test, i128, stmt_expr_attributes)]
```
Comment (goodmanjonathan) [inline src/test/run-pass/saturating-float-casts.rs]: Stray `feature(i128)` left behind.

File: src/test/run-pass/u128-as-f32.rs [hunk 7]
```diff
@@ -10,7 +10,7 @@
 
 // ignore-emscripten u128 not supported
 
-#![feature(test, i128, i128_type)]
+#![feature(test, i128)]
```
Comment (goodmanjonathan) [inline src/test/run-pass/u128-as-f32.rs]: Stray `feature(i128)` left behind.

File: src/test/ui/error-codes/E0658.stderr (hunk 4 above)
Comment (goodmanjonathan) [reply src/test/ui/error-codes/E0658.stderr]: Er, shouldn't this test be testing another feature now :P

**Comment by goodmanjonathan** [comment]
Regarding the first tidy failure, it looks like you forgot to change a few items to `#[stable]` starting on line 3961 of that file.

File: src/test/ui/feature-gate-i128_type2.rs [hunk 8]
```diff
@@ -10,20 +10,20 @@
...
```
Comment (est31) [inline src/test/ui/feature-gate-i128_type2.rs]: You'll probably have to remove this line.

File: src/test/ui/error-codes/E0658.stderr (hunk 4 above)
Comment (mark-i-m) [reply src/test/ui/error-codes/E0658.stderr]: Yeah, the whole stderr needs to change... I will fix this one after I get the rest compiling successfully.

File: src/test/ui/feature-gate-i128_type2.rs (hunk 5 above)
Comment (mark-i-m) [reply src/test/ui/feature-gate-i128_type2.rs]: I left this one as a test for the repr128 gate.

Comment (mark-i-m) [reply src/test/ui/feature-gate-i128_type2.rs]: Oh, it looks like there is already a test for that.

File: src/test/ui/feature-gate-i128_type2.rs (hunk 8 above)
Comment (mark-i-m) [reply src/test/ui/feature-gate-i128_type2.rs]: Thanks :) I removed the test altogether since there is already a test for the repr128 gate.

**Comment by mark-i-m** [comment]
Thanks @goodmanjonathan and @PlasmaPower ! I am fixing those things now :)

**Comment by mark-i-m** [comment]
Hmm... I am still getting this tidy error:

```
tidy error: /nobackup/rust/src/libcore/num/mod.rs:3961: mismatches to previous in: ["stability level", "since", "tracking issue"]
```

But I fixed that line?

**Comment by mark-i-m** [comment]
Re: https://github.com/rust-lang/rust/pull/49101#issuecomment-373890564

How do I actually fix this? Every time `./x.py` updates submodules my dirty changes get reverted...

**Comment by PlasmaPower** [comment]
I'd get the change on a different branch or the submodule's repository, then change the submodule's to point to that branch. I'm not sure if that's the right way to do this though.

**Comment by kennytm** [comment]
@mark-i-m Add `[build] submodules = false` to your config.toml to disable automatic submodule update. 

(Using `__git_ps1` is highly recommended to visualize stale submodules after auto-update is disabled.)

**Comment by nagisa** [comment]
I’ll review this after I get some sleep.

**Comment by mark-i-m** [comment]
Travis build is now failing because of the compiler_builtins error. Should I make a PR to that repo?

**Comment by mark-i-m** [comment]
Ok, I think the compiler_builtins error should be the last thing stopping this PR.

**Comment by mark-i-m** [comment]
Made a PR on compiler_builtins

**Comment by mark-i-m** [comment]
https://github.com/rust-lang-nursery/compiler-builtins/pull/235

**Comment by nagisa** [comment]
The changes here LGTM. Ping me when it is ready to land (i.e. the submodule changes have been merged etc).

**Comment by mark-i-m** [comment]
Thanks @nagisa. I added the two blocking PRs to the OP above... compiler_builtins is the only thing keeping this from compiling at the moment...

**Comment by mark-i-m** [comment]
@nagisa rust-lang-nursery/compiler-builtins#235 has merged, and rust-lang/book#1230 is about to.

**Comment by kennytm** [comment]
I think we are supposed to land this *before* the docs changes rust-lang/book#1230.

**Comment by nagisa** [comment]
There are compilation errors that must be fixed before I can r+ this.

**Comment by mark-i-m** [comment]
Hmm... rust-lang-nursery/compiler-builtins#235 was supposed to fix that...

**Comment by mark-i-m** [comment]
ok, that has merged now :)

**Comment by PlasmaPower** [comment]
I think you'll need to update the submodule for this PR again.

**Comment by mark-i-m** [comment]
@nagisa Travis is passing :tada:

**Comment by mark-i-m** [comment]
The book chapter has also merged.

**Comment by mark-i-m** [comment]
Should this be in the release notes?

**Comment by kennytm** [comment]
Tidy failed. Please rebase and fix the stability.

```
"tidy", path: "src/tools/tidy", mode: Libstd, is_ext_tool: false, extra_features: [] } -- 69.126
* 534 error codes
* highest error code: E0909
tidy error: C:\projects\rust\src\libcore\num\mod.rs:104: mismatches to previous in: ["stability level", "since", "tracking issue"]
tidy error: C:\projects\rust\src\libcore\num\mod.rs:4049: mismatches to previous in: ["stability level", "since", "tracking issue"]
tidy error: C:\projects\rust\src\libstd\num.rs:30: mismatches to previous in: ["stability level", "since", "tracking issue"]
tidy error: C:\projects\rust\src\libstd\primitive_docs.rs:754: mismatches to previous in: ["stability level", "since", "tracking issue"]
some tidy checks failed
```

**Comment by alexcrichton** [comment]
The tidy failures are a bit obscure, but I fixed them in https://github.com/rust-lang/rust/pull/49317/commits/bbdb20a558dc5f9c6541c35bc06cab5cf87dd66f#diff-433b91cf9a1834bfaa4ac67c9527169a which didn't get a chance to land unfortunately

**Comment by kennytm** [comment]
The `// error-pattern: panicked` is to fix another PR. 

The comments should be kept, but don't use `#[unstable` to avoid confusing tidy.

**Comment by est31** [comment]
Yeah, tidy doesn't do proper parsing, it just works based on a string searching method :). And yet the stuff it found was legitimately outdated :), except for the `since="1.26.0"`.

**Comment by mark-i-m** [comment]
Thanks @kennytm and @alexcrichton ! fixed.

**Comment by mark-i-m** [comment]
ugh... I still missed some. Plus there are a bunch of these:

```
DIR/replace_consts.rs:11:17
[01:26:40]    |
[01:26:40] 11 |     { let foo = ATOMIC_BOOL_INIT; };
[01:26:40]    |                 ^^^^^^^^^^^^^^^^ help: try this: `AtomicBool::new(false)`
[01:26:40]    |
[01:26:40] note: lint level defined here
[01:26:40]   --> $DIR/replace_consts.rs:3:9
[01:26:40]    |
[01:26:40] 3  | #![deny(replace_consts)]
[01:26:40]    |         ^^^^^^^^^^^^^^
```

**Comment by mark-i-m** [comment]
I will try to fix these today or tomorrow, but things are a bit busy for me, so this might not make 1.26 :/

**Comment by mark-i-m** [comment]
Oh hmm... these failures are in clippy, not this repo. Do we have to do the same submodule dance?

**Comment by mark-i-m** [comment]
Ok, I opened https://github.com/rust-lang-nursery/rust-clippy/pull/2574 ... I'm not sure if that's the best way to fix it, but...

**Comment by cramertj** [comment]
Looks like MIRI will also need a patch

**Comment by oli-obk** [comment]
Clippy PR has been merged. again, I don't think it's necessary for miri, because it was broken before.

**Comment by mark-i-m** [comment]
@nagisa Travis is passing again. nth try is the charm?

**Comment by Tom-Phinney** [comment]
FWIW, here are additional files in rustc that as of 2018.05.12 did not handle usize == u128, which will be needed by RISC-V with RV128I option when LLVM or some other backend supports it. The grep string (between « and ») was «#\[cfg\(target_pointer_width = "(16|32|64|128)"\)\]»
- lib_alloc/vec_deque.rs 
- libcore/fmt/num.rs
- libcore/iter/range.rs
- libcore/num/mod.rs
- libcore/num/wrapping.rs
- libcore/tests/hash/sip.re
- libcore/tests/iter.re
- libcore/tests/mem.re
- librustc_data_structures/fx.rs
- libserialize/leb128.rs 
- sys_common/backtrace.rs
- test/compile-fail/huge-enum.rs
- test/compile-fail/issue-17913.rs
- test/compile-fail/int-exceeding-bitshifts2.rs
- test/run-pass/const-negation.rs
- test/run-pass/enum-discrim-autosizing.rs
- test/run-pass/extern-types-pointer-cast.rs
- test/run-pass/huge-largest-array.rs
- test/run-pass/issue-2895.rs
- test/run-pass/num-wrapping.rs
- test/run-pass/struct-return.rs
- test/run-pass/vec-fixed-length.rs
- test/run-pass/wrapping_int_api.rs

**Comment by est31** [comment]
@Tom-Phinney it would be best to [file an issue](https://github.com/rust-lang/rust/issues/new) about this as inside this PR thread this is likely going to get lost.

**Pull request merged**